feat: setup for registration servicers

reginawang3495 opened the pull request:
## Summary
- setup protos for use in making `cloudRegistration` servicer and `registration servicer`
- setup storage
- setup constants needed
<!-- Enumerate changes you made and why you made them -->

## Test Plan
- added tests
- run all orc8r tests
<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ ] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

pshelar committed:
fix(pipelined): Handle missing proxy port (#10366)

In some cases ifup fails to add the proxy port to
OVS brige, that could result in turning of HE feature as whole.
Following PR adds fallback to handle such cases.

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

electronjoe committed:
fix(feg): address linting error on numVectors size comparison (#10416)

Fixes `feg-workflow` CI on master which presently blocks new PRs. Regression introduced in #10278, in which for some reason `feg-workflow` was not executed by CI.

Linting notes that `uint64` cannot be negative so the first half of this if condition is a no-op.

While we are here, use Golang built-in for maximum `uint32` size comparison instead of magic value.

Signed-off-by: Scott Moeller <electronjoe@gmail.com>

VinayKumar1591 committed:
test(amf): UT for the service request procedure (#10381)

* test(amf): UT for the service request procedure

Signed-off-by: vinay <vinay.tokala@radisys.com>

ganeshg87 committed:
fix(amf): Unit test cases for PDUSession Establishment/Release (#10236)

* fix(amf): Unit test cases for PDUSession Establishment/Release

Test cases:
  -verified memory leaks & code coverage for below Message Handling
    PDU Session Establishment Request Handling by AMF
    PDU Session Release Handling by AMF
        ngap - amf
	amf  - Pipelined
	amf  - smf

Signed-off-by: ganeshg87 <ganesh.gedela@wavelabs.ai>

Neudrino committed:
docs: Hint on Windows as development OS (#10284)

* docs: Link developer guide from prerequisits

Signed-off-by: Fritz Lehnert <Fritz.Lehnert@tngtech.com>

* docs: restrictions to operating system apply

Signed-off-by: Fritz Lehnert <Fritz.Lehnert@tngtech.com>

* Update docs/readmes/basics/prerequisites.md

Co-authored-by: Hunter Gatewood <hgatewood@gmail.com>

* Update docs/readmes/basics/prerequisites.md

Co-authored-by: Hunter Gatewood <hgatewood@gmail.com>

* Update docs/readmes/basics/prerequisites.md

Co-authored-by: Hunter Gatewood <hgatewood@gmail.com>

Co-authored-by: Marie <37634144+themarwhal@users.noreply.github.com>
Co-authored-by: Hunter Gatewood <hgatewood@gmail.com>

andreilee committed:
fix(ci): Fix Testim reports success count (#10398)

Signed-off-by: Andrei Lee <andreilee@fb.com>

milapjoshi committed:
fix(ci): refresh firebase token and check valid to start test for specific build (#10287)

themarwhal committed:
feat: build sentry statically with Bazel (#10149)

Signed-off-by: Marie Bremner <marwhal@fb.com>

GANESH-WAVELABS committed:
feat(feg): N7 Proxy Service (#10360)

Neudrino committed:
chore(ci): Github composite action docker image builds (#10378)

themarwhal committed:
chore: move .bzl files into bazel/ (#10394)

Signed-off-by: Marie Bremner <marwhal@fb.com>

pshelar committed:
fix(pipelined): Fix UT breakage (#10449)

This commit somehow did not made it into PR #10366

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

Co-authored-by: Pravin B Shelar <pravin.ovn@gmail.com>

hcgatewood committed:
chore(orc8r): Provide better error message when bootstrap challenge fails (#10445)

Signed-off-by: Hunter Gatewood <hgatewood@gmail.com>

koolzz committed:
chore(pipelined): Re-enable check quota pipelined unit test (#10181)

Signed-off-by: Nick Yurchenko <koolzz@fb.com>

talkhasib committed:
feat(ci): Add WL_LAB_5G test worker to list of workers (#10457)

Signed-off-by: Tariq Al-Khasib <talkhasib@fb.com>

ulaskozat committed:
test(mme): Add unit test for idle active transition (#10447)

Signed-off-by: Ulas Kozat <kozat@fb.com>

sanchezfdezjavier committed:
docs: Adds installation support for Ubuntu (#10413)

Co-authored-by: Javi Sanchez <javi@Javis-MacBook-Pro.local>

reginawang3495 committed:
feat: setup for registration SQUASHED

Signed-off-by: Regina Wang <rwang3495@gmail.com>

koolzz committed:
chore(pipelined): Add pipelined rpc unit test (#10385)

Signed-off-by: Nick Yurchenko <koolzz@fb.com>

Neudrino committed:
fix(ci): GHCR access restrictions from GITHUB_TOKEN in new CI (#10460)

... make reactivation of the older CI necessary until the access is granted.

This reverts commit 513485801116a63f4ab6b6f0b5d8c88357223e16.
This partly reverts commit 9a82f48ae333632819b74c511b92db08d2078a38.

Signed-off-by: Fritz Lehnert <Fritz.Lehnert@tngtech.com>

ardzoht committed:
fix(agw): Fixing py redis library conflict due to version (#10462)

Signed-off-by: Alex Rodriguez <alexrod@fb.com>

jwagantall committed:
refactor: Add AMIS GitHub Actions Workflow (#9578)

Add initial AMIS GHA workflow.
These workflow is migrated from CircleCI.

Signed-off-by: Jessica Wagantall <jwagantall@linuxfoundation.org>

GANESH-WAVELABS committed:
feat(feg): N7 Proxy service metrics and health check implementation (#10362)

andreilee committed:
docs(nms): Add back multitenancy setup doc (#10467)

themarwhal committed:
feat: Migrate CWAG sessiond build to be built with Bazel (#10204)

Signed-off-by: Marie Bremner <marwhal@fb.com>

sreedharkumartn committed:
fix(amf): support nssai, dnn parameters in PDU Session Establishment (#9975)

Signed-off-by: sreedharkumartn <sreedhar.kumar@wavelabs.ai>

reginawang3495 committed:
add tests and new proto info

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
remove space

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
add back go.mod

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
add some other files

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
self code review

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
add error logging

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
self code review part 2

Signed-off-by: Regina Wang <rwang3495@gmail.com>

ardzoht committed:
test(mme): Cover s1ap mme handlers tests (#10329)

Signed-off-by: Alex Rodriguez <alexrod@fb.com>

panyogesh committed:
feat(amf): Slice type name and type change in amf (#10392)

Fix:
  1. Change the name of Slice Type and Slice Differentiator
     and appending amf in front.
  2. Changing the type of slice to int as per spec.
  3. Removing prefix 0x from default values of slice
     as per the spec.

Test:
  1. Testing of SD mismatch case with Simulator.
  2. Unit testcases.

Signed-off-by: Yogesh Pandey <yogesh@wavelabs.ai>

alexzurbonsen committed:
chore(agw): Updating Python Sentry SDK to v1.5.0 (#10195)

Updated Sentry SDK version in setup.py of gateway and orc8r.

Signed-off-by: Alexander zur Bonsen <alexander.zur.bonsen@tngtech.com>

alexzurbonsen committed:
chore(smsd): enable selective opt-in logging to Sentry for smsd (#10099)

Send errors to Sentry if MME communication fails.

Signed-off-by: Alexander zur Bonsen <alexander.zur.bonsen@tngtech.com>

quentinDERORY committed:
fix(ci): Only trigger integ tests on push and not on PR (#10483)

Signed-off-by: quentinDERORY <15911421+quentinDERORY@users.noreply.github.com>

quentinDERORY committed:
fix(ci): Bad credentials in deploy on PR (#10482)

Signed-off-by: quentinDERORY <15911421+quentinDERORY@users.noreply.github.com>

reginawang3495 committed:
feat(orc8r): add control_proxy.yml endpoints (#10353)

hcgatewood commented on orc8r/cloud/go/services/bootstrapper/const.go:
```diff
@@ -13,4 +13,14 @@ limitations under the License.
+const (
+	ServiceName  = "BOOTSTRAPPER"
+	DBTableName  = "bootstrapper"
+	TokenPrepend = "reg-"
```
```suggestion
	TokenPrepend = "reg_"
```

hcgatewood commented on orc8r/cloud/go/services/bootstrapper/const.go:
```diff
@@ -13,4 +13,14 @@ limitations under the License.
-const ServiceName = "BOOTSTRAPPER"
+const (
+	ServiceName  = "BOOTSTRAPPER"
+	DBTableName  = "bootstrapper"
```
```suggestion
	BlobstoreTableName  = "bootstrapper"
```

And fix throughout -- the fact that blobstore happens to back to a db is an implementation detail

hcgatewood commented on orc8r/cloud/go/services/bootstrapper/servicers/registration/storage.go:
```diff
@@ -0,0 +1,168 @@
+	"magma/orc8r/cloud/go/blobstore"
+	"magma/orc8r/cloud/go/services/bootstrapper"
+	"magma/orc8r/cloud/go/storage"
+	mErrors "magma/orc8r/lib/go/errors"
```
```suggestion
	merrors "magma/orc8r/lib/go/errors"
```

hcgatewood commented on orc8r/cloud/go/services/bootstrapper/servicers/registration/storage.go:
```diff
@@ -0,0 +1,168 @@
+		return err
+	}
+
+	if oldNonce != "" {
+		oldNonceTK := storage.TKs{{
+			Type: string(bootstrapper.NonceTokenToInfoMap),
+			Key:  oldNonce,
+		}}
+		err = store.Delete(networkWildcard, oldNonceTK)
+		if err != nil {
+			if err != mErrors.ErrNotFound{
+				return err
+			}
+		}
+	}
```
I wouldn't worry about deletion. Or, if you do wanna include it -- add it to a different method

reginawang3495 replied:
Why not delete?

hcgatewood replied:
1. It's just a cache, we can handle cache cleanup separately if it becomes an issue, 2. It feels cleaner to separate them out -- it might feel surprising that a "get token" call would delete things

In either case, if you wanna include the "delete old token", let's just add a separate store method to handle it -- since it doesn't need to be in the same tx as the get/set

hcgatewood commented on orc8r/cloud/go/services/bootstrapper/servicers/registration/storage.go:
```diff
@@ -0,0 +1,168 @@
+	return &tokenInfo, store.Commit()
+}
+
+func (b *blobstoreStore) IsNonceUnique(nonce string) (bool, error) {
```
The probability of a non-unique nonce is trivial. We do wanna check for it, but it can be done in the "set" method rather than having its own method

reginawang3495 replied:
I had a thought for using this with a couple max retries while generating a nonce in “CloudRegistration”

hcgatewood replied:
Makes sense. Given the probability + scale we're working with though, it'll be simpler to just throw an error when it does happen, and caller can retry

This is estimated to never happen in practice 🙂

hcgatewood commented on orc8r/cloud/go/services/bootstrapper/servicers/registration/storage.go:
```diff
@@ -0,0 +1,168 @@
+	if err != nil {
+		return blobstore.Blob{}, errors.Wrap(err, "Error marshaling protobuf")
+	}
+	return blobstore.Blob{
+		Type:  string(blobType),
+		Key:   key,
+		Value: marshaledTokenInfo,
+	}, nil
```
Nit: create the obj first, then return it. Looks slightly cleaner, but also easier for debug printing

hcgatewood commented on orc8r/cloud/go/services/bootstrapper/servicers/registration/storage_test.go:
```diff
@@ -0,0 +1,114 @@
+	"magma/orc8r/lib/go/protos"
+)
+
+var (
+	unusedNonce = "unusedNonce"
+
+	gatewayPreregisterInfo = protos.GatewayPreregisterInfo{
+		NetworkId: "networkID1",
+		LogicalId: "logicalID1",
+	}
+
+	tokenInfo1 = protos.TokenInfo{
+		GatewayPreregisterInfo: &gatewayPreregisterInfo,
+		Nonce:                  "someNonce",
+		Timeout:                nil,
+	}
+
+	tokenInfo2 = protos.TokenInfo{
+		GatewayPreregisterInfo: &gatewayPreregisterInfo,
+		Nonce:                  "someNonce2",
+		Timeout:                nil,
+	}
+)
```
Move this decl inside the test

hcgatewood commented on orc8r/cloud/go/services/bootstrapper/servicers/registration/storage_test.go:
```diff
@@ -0,0 +1,114 @@
+)
+
+func TestBlobstoreStore(t *testing.T) {
+	db, err := sqorc.Open("sqlite3", ":memory:")
+	assert.NoError(t, err)
+	factory := blobstore.NewSQLStoreFactory(bootstrapper.DBTableName, db, sqorc.GetSqlBuilder())
+	assert.NoError(t, factory.InitializeFactory())
```
https://github.com/magma/magma/blob/2104a1c17c4a773dca84d5397665e2db89d565a2/orc8r/cloud/go/test_utils/test_db.go#L49-L58

hcgatewood commented on orc8r/cloud/go/services/bootstrapper/servicers/registration/storage_test.go:
```diff
@@ -0,0 +1,114 @@
+
+	tokenInfo, err = s.GetTokenInfoFromNonce(tokenInfo2.Nonce)
+	assert.NoError(t, err)
+	assert.Equal(t, tokenInfo2, *tokenInfo)
```
Nit: rather than taking the value of the "got" objs, instead make the "want" objs pointers themselves in their definitions (above)

reginawang3495 replied:
ok FYI I had to get values in other places because of this

hcgatewood commented on orc8r/protos/bootstrapper.proto:
```diff
@@ -71,3 +73,82 @@ service Bootstrapper {
+  rpc GetToken(GetTokenRequest) returns (GetTokenResponse) {}
+
+  // GetInfoForGatewayRegistration sends out information needed for gateways to register themselves
+  rpc GetInfoForGatewayRegistration(GetInfoForGatewayRegistrationRequest) returns (GetInfoForGatewayRegistrationResponse) {}
```
"GetGatewayRegistrationInfo"

hcgatewood commented on orc8r/protos/bootstrapper.proto:
```diff
@@ -71,3 +73,82 @@ service Bootstrapper {
+
+  // GetGatewayPreregisterInfo sends out information that has been set when the operator preregistered
+  // the device with networkID and logicalID
+  rpc GetGatewayPreregisterInfo(GetGatewayPreregisterInfoRequest) returns (GetGatewayPreregisterInfoResponse) {}
```
"GetGatewayDeviceInfo"

hcgatewood reviewed (changes requested):
Looks great! Couple small things 🙂

jwagantall committed:
fix: Update BASE_REFERENCE in AMIS workflow (#10490)

Set BASE_REFERENCE to "master" for schedule triggered
runs since github.base_ref is not available.
Use github.base_ref for workflow triggers.

Signed-off-by: Jessica Wagantall <jwagantall@linuxfoundation.org>

quentinDERORY committed:
fix(ci): Permit workflow to continue if other steps are failing (#10485)

Signed-off-by: quentinDERORY <15911421+quentinDERORY@users.noreply.github.com>

prabinakpattnaik committed:
feat(pipelined): Scalability test for 5g call flow in UPF (#9837)

Signed-off-by: mehul jindal <mehul.jindal@wavelabs.ai>

Co-authored-by: mehul jindal <mehul.jindal@wavelabs.ai>

themarwhal committed:
chore: silence gazelle warnings seen with Bazel build (#10456)

Signed-off-by: GitHub <noreply@github.com>

reginawang3495 committed:
address code review

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
fix tests

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
add back go,mod

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
more comments and stuff

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
code review wo gnerate

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
push all generate stuff

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
add timeout remove nil

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
address code review

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
fix tests

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
fix tests

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
fix test bug

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
add to description without generate

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
generate

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
add more comments and other small things

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
generate

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
generate and add go.mod

Signed-off-by: Regina Wang <rwang3495@gmail.com>

reginawang3495 committed:
captilize a letter

Signed-off-by: Regina Wang <rwang3495@gmail.com>
